chains/evm/contracts/pools/FastTransferTokenPoolAbstract.sol

```diff
@@ -0,0 +1,407 @@
+  function fastFill(
+    bytes32 fillRequestId,
+    uint64 sourceChainSelector,
+    uint256 srcAmount,
+    uint8 srcDecimals,
+    address receiver
+  ) public virtual {
+    bytes32 fillId = keccak256(abi.encodePacked(fillRequestId, srcAmount, receiver));
+    address filler = s_fills[fillId];
+    if (filler != address(0)) revert AlreadyFilled(fillRequestId);
+
+    {
+      LaneConfig storage laneConfig = s_fastTransferLaneConfig[sourceChainSelector];
+      if (laneConfig.fillerAllowlistEnabled) {
+        if (!laneConfig.fillerAllowList[msg.sender]) revert FillerNotWhitelisted(sourceChainSelector, msg.sender);
+      }
+    }
+    // Transfer tokens from filler to receiver
+    uint256 destAmount = _transferFromFiller(sourceChainSelector, msg.sender, receiver, srcAmount, srcDecimals);
+    // Record fill
+    s_fills[fillId] = msg.sender;
+    emit FastFill(fillRequestId, fillId, msg.sender, destAmount, receiver);
+  }
+
+  // @inheritdoc CCIPReceiver
+  function _ccipReceive(
+    Client.Any2EVMMessage memory message
+  ) internal override onlyRouter {
+    // Decode message data directly into variables
+    MintMessage memory mintMessage = abi.decode(message.data, (MintMessage));
+    {
+      _settle(
+        message.sourceChainSelector,
+        message.messageId,
+        message.sender,
+        mintMessage.srcAmountToTransfer,
+        mintMessage.srcDecimals,
+        mintMessage.fastTransferFee,
+        address(uint160(uint256(bytes32(mintMessage.receiver))))
+      );
+    }
+    emit FastFillSettled(message.messageId);
+  }
+
+  /// @notice Handles the token to transfer on fast fill request at source chain
+  /// @param destinationChainSelector The destination chain selector
+  /// @param sender The sender address
+  /// @param amount The amount to transfer
+  function _handleTokenToTransfer(uint64 destinationChainSelector, address sender, uint256 amount) internal virtual;
+
+  /// @notice Transfers tokens from the filler to the receiver
+  /// @param sourceChainSelector The source chain selector
+  /// @param filler The address of the filler
+  /// @param receiver The address of the receiver
+  /// @param srcAmount The amount to transfer
+  /// @param srcDecimals The decimals of the source token
+  /// @return destAmount The amount transferred to the receiver on the destination chain
+  function _transferFromFiller(
+    uint64 sourceChainSelector,
+    address filler,
+    address receiver,
+    uint256 srcAmount,
+    uint8 srcDecimals
+  ) internal virtual returns (uint256 destAmount);
+
+  function _settle(
+    uint64 sourceChainSelector,
+    bytes32 fillRequestId,
+    bytes memory sourcePoolAddress,
+    uint256 srcAmount,
+    uint8 srcDecimal,
+    uint256 fastTransferFee,
+    address receiver
+  ) internal virtual {
+    _validateSettlement(sourceChainSelector, sourcePoolAddress);
+
+    // Common calculations
+    uint256 localAmount = _calculateLocalAmount(srcAmount, srcDecimal);
```

Comment:
1. Please add more descriptive comments to indicate that these two amounts are for the source amount and the fee amount. 
2. Can the fee token and the send token be different? If so is it safe to use the same srcDecimal for both?